src/peergos/server/space/JdbcUsageStore.java

```diff
@@ -0,0 +1,370 @@
+
+    public JdbcUsageStore(Connection conn, SqlSupplier commands) {
+        this.conn = conn;
+        this.commands = commands;
+        init(commands);
+    }
+
+    private synchronized void init(SqlSupplier commands) {
+        if (isClosed)
+            return;
+
+        try {
+            commands.createTable(commands.createUsageTablesCommand(), conn);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void initialized() {
+        // TODO can we remove this method?
+    }
+
+    @Override
+    public void addUserIfAbsent(String username) {
+        try (PreparedStatement insert = conn.prepareStatement(commands.insertOrIgnoreCommand("INSERT ", "INTO users (name) VALUES(?)"))) {
+            insert.setString(1, username);
+            insert.executeUpdate();
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+        int userId = getUserId(username);
+        try (PreparedStatement insert = conn.prepareStatement(commands.insertOrIgnoreCommand("INSERT ", "INTO userusage (user_id, total_bytes, errored) VALUES(?, ?, ?)"))) {
+            insert.setInt(1, userId);
+            insert.setLong(2, 0);
+            insert.setBoolean(3, false);
+            insert.executeUpdate();
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    private long getSpaceUsed(int userId) {
+        try (PreparedStatement insert = conn.prepareStatement("SELECT total_bytes FROM userusage WHERE user_id = ?;")) {
+            insert.setInt(1, userId);
+            ResultSet resultSet = insert.executeQuery();
+            resultSet.next();
+            return resultSet.getLong(1);
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    @Override
+    public void confirmUsage(String username, PublicKeyHash writer, long usageDelta) {
+        int userId = getUserId(username);
+        long currentUsage = getSpaceUsed(userId);
+        try (PreparedStatement insert = conn.prepareStatement("UPDATE userusage SET total_bytes = ?, errored = ? " +
+                "WHERE user_id = ? AND total_bytes = ?;")) {
+            insert.setLong(1, currentUsage + usageDelta);
+            insert.setBoolean(2, false);
+            insert.setInt(3, userId);
+            insert.setLong(4, currentUsage);
+            int count = insert.executeUpdate();
+            if (count != 1)
+                throw new IllegalStateException("Didn't update one record!");
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+        clearPendingUsage(username, writer);
```

Comment:
the call to clearPendingUsage should be part of a db transaction

Reply:
In theory yes, but I think it doesn't matter if this races with a addPendingSpace call, because the result will be corrected when confirming the transaction anyway. What do you think?